sql-plugin/src/main/scala/org/apache/spark/rapids/hybrid/RapidsHostBatchProducer.scala

```diff
@@ -0,0 +1,235 @@
+  @volatile private var writeIndex: Int = 0
+
+  // This lock guarantees that the status change of isProducing and the check of isProducing
+  // can NOT be executed currently.
+  private lazy val isInProgressLock = new ReentrantLock()
+  // Asleep the consumerThread if the buffer is empty.
+  // Awake the consumerThread as soon as the next element being ready.
+  private lazy val emptyLock = new ReentrantLock()
+  // Asleep the producerThread if the buffer is full.
+  // Awake the producerThread as soon as elements being consumed.
+  private lazy val fullLock = new ReentrantLock()
+
+  private var producerThread: Thread = _
+
+  private lazy val buffer: Array[Either[Throwable, Array[RapidsHostColumn]]] = {
+    Array.ofDim[Either[Throwable, Array[RapidsHostColumn]]](capacity)
+  }
+
+  private lazy val produceFn: Runnable = new Runnable {
+
+    // This context will be got in the main Thread during the initialization of `produceFn`
+    private val taskContext: TaskContext = TaskContext.get()
+
+    override def run(): Unit = {
+      TrampolineUtil.setTaskContext(taskContext)
+      isInProgressLock.lock()
+      try {
+        do {
+          isProducing = true
+          isInProgressLock.unlock()
+          // Wait for empty slot of RingBuffer
+          do {
+            fullLock.synchronized {
+              if (writeIndex - readIndex == capacity) {
+                fullLock.wait()
+              }
+            }
+          } while (writeIndex - readIndex == capacity)
+          // Trigger the base iterator
+          buffer(writeIndex % capacity) = Right(base.next())
+          // Update the writeIndex and try to awake the consumer thread which calls waitForNext
+          emptyLock.synchronized {
+            writeIndex += 1
+            emptyLock.notify()
+          }
+
+          isInProgressLock.lock()
+          isProducing = false
+          logInfo(s"[$taskAttId] PreloadedIterator produced $writeIndex batches, " +
+            s"currently preloaded batchNum: ${writeIndex - readIndex}"
+          )
+        }
+        while (base.hasNext)
+        isInProgressLock.unlock()
+      } catch {
+        case ex: Throwable =>
+          // transfer the exception info to the main thread as an interrupted signal
+          buffer(writeIndex % capacity) = Left(ex)
+          writeIndex += 1
+          isProducing = false
+          if (isInProgressLock.isHeldByCurrentThread) {
+            isInProgressLock.unlock()
+          }
+          throw new RuntimeException(ex)
+      } finally {
+        TrampolineUtil.unsetTaskContext()
+      }
+    }
+  }
+
+  override def hasNext: Boolean = {
+    // Lazy Init of the actual producer
+    if (!isInit) {
+      withResource(new NvtxWithMetrics("waitForCPU", NvtxColor.RED, waitTimeMetric)) { _ =>
+        if (!base.hasNext) {
+          return false
+        }
+        isInit = true
+        isProducing = true
+        producerThread = new Thread(produceFn)
```

Comment:
Pass in a thread name, then stack dumping will contain this thread name.

Reply:
Done